File: books/food_lines/food.py
def solve(n, m, original_lines: tuple):
    lines = list(original_lines)
    joined = [] # To track the lenght of the lines we joing, for the solution
    for _ in range(m):
        # Get the shortest line information
        min_value = min(lines)
        min_index = lines.index(min_value)

        # Note the length of the line we joined for the solution
        joined.append(min_value)

        # Joined line is now one longer
        lines[min_index] += 1
    return joined

test_cases = {
    (3, 4, (3, 2, 5)): [2, 3, 3, 4],
}
for case, expected in test_cases.items():
    n, m, lines = case
    result = solve(n, m, lines)
    assert result == expected, f"Failed: solve{input} = {expected}, got {result}"
print("passed all test cases")

File: books/food_lines/food.c
#include <stdio.h>
#include <stdlib.h>

// returns the index of the minimum value, -1 if not valid input
int find_min(int arr[], int len_arr) {
    if ( length(arr) == 0 ) {
        return -1;
    }
    int min_index = 0;
    int min_value = arr[0];
    // Start at 1 since index 0 is necessarily the min starting out
    for( int i = 1; i < len_arr; i++ ) {
        if ( arr[i] < min_value ) {
            min_index = i;
            min_value = arr[i];
        }
    }
    return min_index;
}

int* solve(int n, int m, int lines[]) {
    int joined_lines[m];
    int min_index = -1;
    for ( int i = 0; i < m; i++ ) {
        min_index = find_min(lines, n);
        if ( min_index != -1 ) {
            // Record the smallest line to join
            joined_lines[i] = lines[min_index];

            // Add the person that joined that line
            lines[min_index] += 1;
        }
        else {
            printf("Error");
        }
    }
}

int main() {
    solve(3, 4, [2, 3, 5]);
}